README.md

# react-native-text-gradient
[![npm version](https://badge.fury.io/js/react-native-text-gradient.svg?t=1495378566925)](https://badge.fury.io/js/react-native-text-gradient)
[![Dependency Status](https://david-dm.org/iyegoroff/react-native-text-gradient.svg?t=1495378566925)](https://david-dm.org/iyegoroff/react-native-text-gradient)
[![typings included](https://img.shields.io/badge/typings-included-brightgreen.svg?t=1495378566925)](#typescript)
[![npm](https://img.shields.io/npm/l/express.svg?t=1495378566925)](https://www.npmjs.com/package/react-native-text-gradient)

React-Native text gradient component for iOS & Android.

## Getting started

`$ npm install react-native-text-gradient --save`

### Mostly automatic installation

`$ react-native link react-native-text-gradient`

### Manual installation

[link](manual_installation.md)

## Status

- iOS - component works as drop-in replacement for standard `Text` component, e.g. it is possible to have nested gradients;
- Android - currently only basic 'wrapper'-like behavior without nesting is supported, [WIP](https://github.com/iyegoroff/react-native-text-gradient/tree/android-nested-gradient);
- React-Native:
  - with rn >= 0.56.0 use latest version and [patch](#usage-with-rn--0560);
  - with rn >= 0.55.0 use 0.0.9;
  - with rn >= 0.54.0 use 0.0.7;
  - with rn >= 0.53.1 use 0.0.4;
  - rn 0.53.0 is not supported;
  - with rn >= 0.50.0 use 0.0.3.


## Example

```javascript
import { LinearTextGradient } from 'react-native-text-gradient';

<LinearTextGradient
  style={{ fontWeight: 'bold', fontSize: 72 }}
  locations={[0, 1]}
  colors={['red', 'blue']}
  start={{ x: 0, y: 0 }}
  end={{ x: 1, y: 0 }}
>
  THIS IS TEXT GRADIENT
</LinearTextGradient>
```

iOS                                            |  Android
:---------------------------------------------:|:---------------------------------------------:
<img src="img/ios.png" align="left" height="275">  |  <img src="img/android.jpg" align="right" height="275">


## Usage

### LinearTextGradient
Interface is similar to `Text` & [LinearGradient](https://github.com/react-native-community/react-native-linear-gradient)

#### colors
An array of at least two color values that represent gradient colors. Example: `['red', 'blue']` sets gradient from red to blue.
  
#### start
An optional object of the following type: `{ x: number, y: number }`. Coordinates declare the position that the gradient starts at, as a fraction of the overall size of the gradient, starting from the top left corner. Example: `{ x: 0.1, y: 0.1 }` means that the gradient will start 10% from the top and 10% from the left.
 
#### end
Same as start, but for the end of the gradient.
 
#### locations
An optional array of numbers defining the location of each gradient color stop, mapping to the color with the same index in `colors` prop. Example: `[0.1, 0.75, 1]` means that first color will take 0% - 10%, second color will take 10% - 75% and finally third color will occupy 75% - 100%.

#### useViewFrame
Optional. If true gradient will be calculated for text view background frame rather than text frame.

```javascript
<LinearTextGradient
  numberOfLines={1}
  useViewFrame={true}
  locations={[0.5, 0.95]}
  // note colors like '#FF000000' are used for seamless transition to transparent
  colors={['#FF0000', '#FF000000']}
  start={{ x: 0, y: 0 }}
  end={{ x: 1, y: 0 }}
>
  %%%%%%%%%%%%%%%%%%%%%%
</LinearTextGradient>
```

<img src="img/useViewFrame.png" width="300">


## Usage with rn >= 0.56.0

Wait until https://github.com/facebook/react/pull/13211 will be merged or patch react-native to remove failing invariant checks

`$ node node_modules/react-native-text-gradient/patch-rn.js`


## Caveats

When mixing several text gradients and `Text`s top component always should be text gradient.
```javascript
<LinearTextGradient {...someGradientProps}>
  <Text>123</Text>
  qwerty
  <LinearTextGradient {...anotherGradientProps}>321</LinearTextGradient>
</LinearTextGradient>
```
This is necessary because only top text component is 'mapped' to actual native node and its children are 'virtual' from the native perspective.

## FAQ

#### Is it ready for production?
- Yes, I use it in production


RELEASE_NOTES.md
[release notes](https://github.com/iyegoroff/react-native-text-gradient/releases)

example/App.js
import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  ScrollView
} from 'react-native';
import {LinearTextGradient} from 'react-native-text-gradient';

export default class App extends Component {
  render() {
    const size = 1;
    return (
      <ScrollView style={styles.container}>
        {/* <LinearTextGradient
          style={styles.gradient}
          locations={[0, 0.5, 1]}
          colors={['red', 'green', 'blue']}
        >
          HORIZONTAL GRADIENT███████████████████████████████████████████████████████████████████████
        </LinearTextGradient>
        <LinearTextGradient
          style={styles.gradient}
          locations={[0, 0.5, 1]}
          colors={['#00FFFF', '#FF00FF', '#FFFF00']}
          end={{ x: 0, y: 1 }}
        >
          VERTICAL GRADIENT█████████████████████████████████████████████████████████████████████████
        </LinearTextGradient> */}
        <LinearTextGradient
          style={styles.gradient}
          locations={[0, 0.5, 1]}
          colors={['yellow', 'purple', 'lime']}
          start={{ x: 1, y: 0 }}
          end={{ x: 0, y: 1 }}
        >
        {/* ██████████████████ */}
          {/* ██████████████████████████████████████████████████████████████████████████████████████████ */}
          {range(0, size).map(x => (
          <LinearTextGradient
            key={x}
            style={styles.gradient}
            useAbsoluteSizes={false}
            locations={[0, 0.25, 0.5, 0.75, 1]}
            colors={['red', 'white', 'green', 'white', 'purple']}
            start={{ x: 0, y: 0 }}
            end={{ x: 0, y: 1 }}
          >
            █████████ qeifmgb fw- wrkbm
          </LinearTextGradient>
          ))}
          {/* {range(0, size).map(x => '█████████')} */}
          {/* ██████████████████████████████████████████████████████████████████████████████████████████ */}
        </LinearTextGradient>
        <View style={styles.view}>
         {range(0, size).map(x => (
           <View key={x} style={styles.subview} />
         ))}
        </View>
      </ScrollView>
    );
  }
}

function range(start, end) {
  var array = new Array();
  for (var i = start; i < end; i++) {
    array.push(i);
  }
  return array;
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 5,
    backgroundColor: 'lightgray'
  },
  gradient: {
    fontSize: 40,
    // lineHeight: 47,
    marginBottom: 20,
    backgroundColor: 'wheat'
  },
  view: {
    position: 'absolute',
    // paddingTop: 5.5
  },
  subview: {
    width: 30,
    height: 47,
    backgroundColor: 'transparent',
    borderColor: 'black',
    borderBottomWidth: 1
  }
});
